src/gordon_gcp/plugins/event_consumer.py

```diff
@@ -50,15 +275,163 @@ class GPSEventConsumer:
+        msg = GEventMessage(pubsub_msg, data)
+        msg.append_to_history(log_entry, self.phase)
+        return msg
+
+    def _get_and_validate_pubsub_msg_schema(self, pubsub_msg_data):
+        for _schema in self._message_schemas:
+            try:
+                self._validator.validate(pubsub_msg_data, _schema)
+                return _schema
+            except exceptions.InvalidMessageError as e:
+                continue
+
+    async def _handle_pubsub_msg(self, pubsub_msg):
+        msg_id = pubsub_msg._ack_id.split(':')[-1]
+        msg_logger = _utils.MessageLogger(
+            self._logger, {'msg_id': msg_id})
+        msg_logger.info('Received new message.')
+
+        try:
+            data = json.loads(pubsub_msg.data)
+            msg_logger.debug(f'Received data: {data}')
+        except json.JSONDecodeError as e:
+            msg = ('Issue loading message data as JSON. '
+                   f'Given data: {pubsub_msg.data}')
+            msg_logger.warn(f'{msg}')
+            pubsub_msg.ack()
+            return
+
+        schema = self._get_and_validate_pubsub_msg_schema(data)
+        if schema is None:
+            msg_logger.warn('Invalid message received, acking.')
+            pubsub_msg.ack()
+            return
+
+        msg_logger.debug(f'Message is valid for "{schema}" schema.')
+
+        event_msg_data = self._parser.parse(data, schema)
+        event_msg = self._create_gevent_msg(
+            pubsub_msg, event_msg_data, schema)
+
+        # if the message has resource records, assume it has all info
+        # it needs to be published, and therefore is already enriched
+        phase = 'enrich' if event_msg_data['resourceRecords'] else 'consume'
+        msg_logger.debug(f'Updating message phase to "{phase}".')
+        await self.update_phase(event_msg, phase=phase)
+
+        msg_logger.debug(f'Adding message to the success channel.')
+        await self.success_channel.put(event_msg)
+
+    def _thread_pubsub_msg(self, pubsub_msg):
+        # Create a thread event
+        event = threading.Event()
+        # create a new thread for the pubsub
+        thread_name = f'GPSThread_msg-id_{pubsub_msg.message_id}'
+        thread = _GPSThread(event, name=thread_name, daemon=True)
+        # start the thread
+        thread.start()
+        # allow the loop's thread to signal to us
+        event.wait()
+        # add a task for the thread (initiates work)
+        thread.add_task(self._handle_pubsub_msg(pubsub_msg))
+        # save thread to stop/clean up during self.cleanup
+        self._threads[pubsub_msg._ack_id] = (event, thread)
+
+    def _manage_subs(self):
+        # TODO (lynn): look into needing flow control, e.g.:
+        # flow_control = pubsub_v1.types.FlowControl(max_messages=10)
+        # self._subscriber.subscribe(self._subscription,
+        #     flow_control=flow_control)
+        # NOTE: automatically extends deadline in the background;
+        #       must `nack()` if can't finish.
```

Comment:
No. I thought about having like a shutdown/plugin cleanup method in which these messages would then be nack'ed. But then I thought: if the process dies, it would eventually (~10 seconds I believe?) be nack'ed since pub/sub wouldn't hear from the process. I can update the comment to make more sense if you'd like.